numba/cuda/cudaimpl.py

```diff
@@ -405,6 +405,116 @@ def ptx_popc(context, builder, sig, args):
+def float16_to_float_cast(context, builder, fromty, toty, val):
+    if fromty.bitwidth == toty.bitwidth:
+        return val
+
+    ty, constraint = float16_float_ty_constraint(toty.bitwidth)
+
+    fnty = ir.FunctionType(context.get_value_type(toty), [ir.IntType(16)])
+    asm = ir.InlineAsm(fnty, f"cvt.{ty}.f16 $0, $1;", f"={constraint},h")
+    return builder.call(asm, [val])
+
+
+@lower_cast(types.Float, types.float16)
+def float_to_float16_cast(context, builder, fromty, toty, val):
+    if fromty.bitwidth == toty.bitwidth:
+        return val
+
+    ty, constraint = float16_float_ty_constraint(fromty.bitwidth)
+
+    fnty = ir.FunctionType(ir.IntType(16), [context.get_value_type(fromty)])
+    asm = ir.InlineAsm(fnty, f"cvt.rn.f16.{ty} $0, $1;", f"=h,{constraint}")
+    return builder.call(asm, [val])
+
+
+def float16_int_constraint(bitwidth):
+    typemap = { 8: 'c', 16: 'h', 32: 'r', 64: 'l' }
+
+    try:
+        return typemap[bitwidth]
+    except KeyError:
+        msg = f"Conversion between float16 and int{bitwidth} unsupported"
+        raise errors.CudaLoweringError(msg)
+
+
+@lower_cast(types.float16, types.Integer)
+def float16_to_integer_cast(context, builder, fromty, toty, val):
+    bitwidth = toty.bitwidth
+    signedness = 's' if toty.signed else 'u'
+    constraint = float16_int_constraint(bitwidth)
+
+    fnty = ir.FunctionType(context.get_value_type(toty), [ir.IntType(16)])
+    asm = ir.InlineAsm(fnty,
+                       f"cvt.rni.{signedness}{bitwidth}.f16 $0, $1;",
+                       f"={constraint},h")
+    return builder.call(asm, [val])
+
+
+@lower_cast(types.Integer, types.float16)
+@lower_cast(types.IntegerLiteral, types.float16)
+def integer_to_float16_cast(context, builder, fromty, toty, val):
+    bitwidth = fromty.bitwidth
+    signedness = 's' if fromty.signed else 'u'
+    constraint = float16_int_constraint(bitwidth)
+
+    fnty = ir.FunctionType(context.get_value_type(toty),
+                           [context.get_value_type(fromty)])
+    asm = ir.InlineAsm(fnty,
+                       f"cvt.rn.f16.{signedness}{bitwidth} $0, $1;",
+                       f"=h,{constraint}")
+    return builder.call(asm, [val])
+
+
+def lower_fp16_binary(fn, op):
+    @lower(fn, types.float16, types.float16)
+    def ptx_fp16_binary(context, builder, sig, args):
+        fnty = ir.FunctionType(ir.IntType(16),
+                               [ir.IntType(16), ir.IntType(16)])
+        asm = ir.InlineAsm(fnty, f'{op}.f16 $0,$1,$2;', '=h,h,h')
+        return builder.call(asm, args)
+
+
+lower_fp16_binary(stubs.fp16.hadd, 'add')
+lower_fp16_binary(stubs.fp16.hsub, 'sub')
+lower_fp16_binary(stubs.fp16.hmul, 'mul')
+
+
+def lower_fp16_unary(fn, op):
+    @lower(fn, types.float16)
+    def ptx_fp16_unary(context, builder, sig, args):
+        if op == 'abs' and cuda.runtime.get_version() < (10, 2):
+            msg = f"CUDA toolkit >= 10.2 required for {op}(float16)"
+            raise errors.CudaLoweringError(msg)
```

Comment:
Following OOB discussion - we don't generally always know what's going to be done with the PTX we produce (e.g. in `compile_ptx`, what the toolkit / driver version used with the output will be) - so rather than erroring, it would be better to provide an alternative implementation for toolkits that don't support `abs.f16`. In this case we'll mask off the signbit, and we'll still use the runtime version to determine which code to generate, because this will result in code generation consistent with the NVVM version currently being used to go from NVVM IR to PTX.

(Also in general we should not introduce errors that are intended to be user-facing in lowering)

Reply:
This is now implemented.

Reply:
Thanks, implementation looks good.